### ISSUE TITLE
magic rerolling pgem quantity question

### ISSUE THREAD

**Issue by zamukan** [kolton/d2bot-with-kolbot#2709]
Hi all
ive tried to find out how bots keep pgems for rerolling without tossing on the ground when theres no crafting main ingredient.

in the current char config set up, its using all pgems except Pskull for crafting and cubing armor wep and shields. 
The char config states that  "// The gems not used by other recipes will be used for magic item rerolling." 
and in my pickit ini, i have disabled all pgems. according to that statement it means it will use only pskulls to reroll magic. 

So i have also tried to keep all pgems with maxquantity 2 and it keeps pgems in the stash and doesnt use for rerolling. 

and the other method ive tried is in the cubing.js i have put pgem numbers such as "561", "586", "5xx" instead of "pgems"
however the bots still drop other pgems if it doesnt have the main ingredient for crafting. 

is there a way around to use all pgems for rerolling without dropping on the ground?

thank you in advance.

**Comment by DarkHorseDre** [comment]
I don't think it will use pskulls to role magic - iirc it either see's pgems and skulls the same (you can check this in cubing.js) or it separates them sometimes for the recipes which require pskulls.

As for your main question:
> the bots still drop other pgems if it doesn't have the main ingredient for crafting.

Yeah this is how it's designed. you'd have to modify the logic to change that.
iirc it waits for the base item then looks for gems - I guess to save space (otherwise you could have a stash full of gems whilst you wait for that level 99 gc :) ).

gems are so common in hell, are you sure this is a big issue?

**Comment by zamukan** [comment]
hi Dre
thanks for the input
yes in my case it does use pskull for reroll magic 
Right. its not a big deal but since the only pgem used for reroll is pskull i wanted to have more chance with rerolling and was thinking kind of miss d2nt style cubing which saves # and type of pgems you choose in stash for cubing and reroll. 

btw i think getting gcs is not an issue if you put 1more logic line it keeps 2gcs in stash and the chance of getting desired mod from reroll is low anyway. 

I gotta leave it as is since im not a programmer.
Thanks Dre

**Comment by DarkHorseDre** [comment]
np - and hey, these are my observations and views - someone with time+ability+understanding of how this works may offer to help ya

I don't remember this setting in d2nt but agree this should all be configurable.. but at least its customisable!

> since the only pgem used for reroll is pskull i wanted to have more chance with rerolling

Wait, are you saying you want to avoid using pskulls in magic reroll recipe?  
if so I'd rewrite your post to make this clear,

```js
this.recipes.push({Ingredients: [Config.Recipes[i][1], "pgem", "pgem", "pgem"], Level: 99, Index: Recipe.Reroll.Magic}); 
```
the recipe requires 1 item to be specific per ingredient, so I'm not sure how to say "561 or 566 or 571 or 576 or 581 or 586" for each of the 3 ingredients.. so if there isn't a way, the functions would need to be modified (it would be a good change, but you gotta get someone to do that)..

I'm asking in discord group - will update if I hear anything..

**Comment by DarkHorseDre** [comment]
ok so this line specifies all the gems to use for rerolling:
```js
gemList = [561, 566, 571, 576, 581, 586, 601],
```

so you could just remove pskulls (601) from that list and it should work:
```js
gemList = [561, 566, 571, 576, 581, 586],
```

I am working on having recipes available but ran out of time - can look tomorrow.
lemme know

**Comment by DarkHorseDre** [comment]
update

**Comment by zamukan** [comment]
Hi Dre
i meant using pskull is fine but when i leave as "pgem" "pgem" "pgem" it only uses pskulls
since i do not know how to configure it uses all pgems for magic reroll,
Ive just put to use Ptopaz, Pdia, Peme 
not all other pgems. 

last week i have also tested with 
for (j = 0; j < this.recipes[i].Ingredients.length; j += **1**) { 
changed 1 to 2 at this line then it keeps all flawless didnt cube 

i dont know what those numbers do at the end of those lines below but ill be testing again with different numbers to see if it keeps pgems even its an ingredient for crafting.
				for (j = 0; j < this.recipes[i].Ingredients.length; j += 1) {
					if (gemList.indexOf(this.recipes[i].Ingredients[j]) > -1) {
						gemList.splice(gemList.indexOf(this.recipes[i].Ingredients[j]), **1**);
		this.gemList = gemList.slice(0);

thanks

**Comment by DarkHorseDre** [comment]
> i meant using pskull is fine but when i leave as "pgem" "pgem" "pgem" it only uses pskulls

This is not default behaviour - pgems uses the gemList which has 7 id's, for each gem and pskulls.  you mustve changed something.

> since i do not know how to configure it uses all pgems for magic reroll,

I dont understand what you are saying - this seems to contradict your last statement?

> Ive just put to use Ptopaz, Pdia, Peme
> not all other pgems.

for what purpose???  shouldnt be necessary. have you read my previous post?
> 
> last week i have also tested with
> for (j = 0; j < this.recipes[i].Ingredients.length; j += **1**) {
> changed 1 to 2 at this line then it keeps all flawless didnt cube
> 
> i dont know what those numbers do at the end of those lines below but ill be testing again with different numbers to see if it keeps pgems even its an ingredient for crafting.
> for (j = 0; j < this.recipes[i].Ingredients.length; j += 1) {
> if (gemList.indexOf(this.recipes[i].Ingredients[j]) > -1) {
> gemList.splice(gemList.indexOf(this.recipes[i].Ingredients[j]), **1**);
> this.gemList = gemList.slice(0);
> 
> thanks

Sounds like you are trying to accomplish many things in this one post and are not being clear in your explanation.

if you want it to collect gems if the roll.magic recipe is enabled but no base item is present, you'll have to change the logic that does this - and thats beyond me!

good luck bro

**Comment by Roycinger** [comment]
> I don't think it will use pskulls to role magic - iirc it either see's pgems and skulls the same (you can check this in cubing.js) or it separates them sometimes for the recipes which require pskulls.
> 
> As for your main question:
> 
> > the bots still drop other pgems if it doesn't have the main ingredient for crafting.
> 
> Yeah this is how it's designed. you'd have to modify the logic to change that.
> iirc it waits for the base item then looks for gems - I guess to save space (otherwise you could have a stash full of gems whilst you wait for that level 99 gc :) ).
> 
> gems are so common in hell, are you sure this is a big issue?

My bot doesnt drop pgems although there is no suitable Grand Charm in the chest. 
To be more clear:
There is no Grand Charm to cube and my stash is getting flooded with pgems.
Is there a way to set the bot to drop pgems unless there's a gc to cube?
I disabled pgems in my pickit and only left `Config.Recipes.push([Recipe.Gem, "Flawless Amethyst"]);` in it (for all kinds of gems.
